File: geometry/coord.c
#include <stdio.h>
#include <math.h>

typedef struct 
{
    double x;
    double y;
} coord_t;

typedef struct 
{
    coord_t rotAt;
    double theta;
    double ofs_x;
    double ofs_y;
} config_t;

typedef coord_t (*converter_t)(coord_t);

coord_t trans(double dx, double dy, coord_t coord){
    coord_t result=coord;
    result.x +=dx;
    result.y +=dy;
    return result;
}

coord_t rotate(double theta, coord_t coord){
    coord_t result = {0,0};
    result.x=cos(theta) * coord.x - sin(theta) * coord.y;
    result.y=sin(theta) * coord.x + cos(theta) * coord.y;
    return result;
}

coord_t trans_by_config(config_t config, coord_t coord){
    return trans(config.ofs_x,config.ofs_y, coord);
}

coord_t rotate_by_config(config_t config, coord_t coord){
    coord_t pre_trans = trans(-config.rotAt.x,-config.rotAt.y, coord);
    coord_t rotated = rotate(config.theta, pre_trans);
    coord_t post_trans =  trans(config.rotAt.x,config.rotAt.y,rotated);
    return post_trans;
}

coord_t convert_by_config(config_t config, coord_t coord){
    return trans_by_config(config, rotate_by_config(config, coord));
}

void map_to_coords(converter_t conv, size_t n, coord_t* in_coord, coord_t* out_coord){
    unsigned int i = 0;
    for (i=0;i<n;i++) out_coord[i] = conv(in_coord[i]);
}

int main(){
    config_t config = { {0.5,0.5}, 3.141592653589793/4,-0.5,-0.5};
    coord_t unit_rect[] = {{0,0},{0,1},{1,1},{1,0}};
    coord_t converted_rect[] = {{0,0}, {0,0}, {0,0},{0,0}};
    {
        unsigned int i=0;
        for (i=0;i<sizeof(unit_rect)/sizeof(unit_rect[0]);i++){
            converted_rect[i]=convert_by_config(config,unit_rect[i]);
        }
    }

    {
        unsigned int i=0;
        for (i=0; i<sizeof(unit_rect)/sizeof(unit_rect[0]);i++){
            printf("(%.6f,%.6f\n",converted_rect[i].x,converted_rect[i].y);
        }
    }
    return 0;
}

File: geometry/coord.js
function clone(obj){
    var f = function(){};
    f.prototype = obj;
    return new f;
}

var trans = function(dx,dy,coord){
    var result = clone(coord);
    result.x += dx;
    result.y += dy;
    return result;
}

var rotate = function(theta,coord){
    var result = clone(coord);
    result.x = Math.cos(theta) * coord.x - Math.sin(theta) * coord.y
    result.y = Math.sin(theta) * coord.x + Math.cos(theta) * coord.y
    return result;
}

var transByConfig = function(config, coord){
    return trans(config.ofsX, config.ofsY, coord);
}

var rotateByConfig = function(config, coord){
    var preTrans = trans(-config.rotAt.x,-config.rotAt.y, coord);
    var rotated = rotate(config.theta, preTrans);
    var postTrans = trans(config.rotAt.x,config.rotAt.y,rotated);
    return postTrans;
}

var convertByConfig = function(config, coord){
    return transByConfig(config, rotateByConfig(config, coord));
}

var config = {
    'rotAt' :{
        'x':0.5,
        'y':0.5,
    },
    'theta':Math.PI/4,
    'ofsX':-0.5,
    'ofsY':-0.5
};

var unit_rect = [
    {'x':0,'y':0},
    {'x':0,'y':1},
    {'x':1,'y':1},
    {'x':1,'y':0}
];

var converted_rect = unit_rect.map(function(coord){
    return convertByConfig(config,coord);
});

converted_rect.map(function(coord){
    console.log('('+coord.x.toFixed(6)+','+coord.y.toFixed(6)+')');
});
